### PR TITLE
Interface with MongoChem server

### PR DESCRIPTION
**Pull request by psavery**
This adds a basic interface with MongoChem server. It is located at "Extensions"->"MongoChem". Currently, it can find public molecules in the database as shown below.

example

There are several hard-coded parts, including the server address to "localhost:8080". But these can be fixed in the future. Additionally, a QTableWidget is used rather than a model/view approach, which is something that should probably be fixed in the future as well...

@cryos

### CODE REVIEW

**Comment by ghutchis** [comment]
While it's not perfect by any means, you might also want to look at the PQR plugin:
https://github.com/OpenChemistry/avogadrolibs/blob/master/avogadro/qtplugins/importpqr/pqrwidget.cpp

I don't know if there's merit to integrating some of these efforts. One case we considered was "look up the thing I just drew" in the server (i.e., exact InChI match).

**Comment by cryos** [comment]
I was just looking at that @ghutchis. This plugin is intended to start with non-authenticated, and then add editing/upload capabilities. I think we should iterate towards some common dialogs for search, preview, etc. I think short term it would be good to experiment in a dedicated plugin before we try and combine things. I would also like to improve the search of other online databases where error reporting is minimal such as PDB, and the structure resolver service.

File: avogadro/qtplugins/mongochem/mongochemwidget.ui [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    <width>620</width>
+    <height>410</height>
+   </size>
+  </property>
+  <property name="windowTitle">
+   <string>Form</string>
+  </property>
+  <layout class="QGridLayout" name="gridLayout">
+   <item row="0" column="0">
+    <widget class="QTableWidget" name="table_molecules">
```
Comment (cryos) [inline avogadro/qtplugins/mongochem/mongochemwidget.ui]: Prefer camel case for names, i.e. tableMolecules

File: avogadro/qtplugins/mongochem/mongochemwidget.ui [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+     </column>
+     <column>
+      <property name="text">
+       <string>InChIKey</string>
+      </property>
+     </column>
+    </widget>
+   </item>
+   <item row="1" column="0">
+    <widget class="QPushButton" name="push_search">
```
Comment (cryos) [inline avogadro/qtplugins/mongochem/mongochemwidget.ui]: As above

**Comment by psavery** [comment]
There's now a config button in the bottom right corner that opens a dialog where the user can set the url and the api key. The line edit for the api key uses a password echo mode. It is then stored as a plain text variable in the program.

The api key is currently not being used for anything. But soon, after the user accepts the config, the api key will be used for authentication, and the resulting token will be stored and used for the http requests.

update

config

**Comment by psavery** [comment]
Molecules can now be downloaded by selecting a row and clicking "Download" at the bottom.

img

**Comment by cryos** [comment]
I think this looks close, when entering the server address we should append the /api/v1 for the user, so they can just paste in the URL, i.e. http://localhost:8080 or https://data.openchemistry.org, it might be good to have a little debug code in to tell the user we are polling for a result on the command line. If we can make it remember the shorter URL and the API key then I  think this is in a place where  it should be merged.

There is a lot of code to submit, the fact it works is great and we can consider possible API updates later.

**Comment by psavery** [comment]
@cryos Thank you! And sounds good, I will work on those.

After the geometry branches in mongochemserver get merged, I'm planning to have Avogadro2 always upload its molecule as a geometry for the calculation, and then run the calculation on that geometry. That way, it will always perform the calculation on whatever is drawn (rather than what is already in the database).

**Comment by psavery** [comment]
Done and done.

**Comment by cryos** [comment]
Let's add the QSettings save for URL and API key (I don't see that), and then merge what we have. Agreed on the geometry, but what we have here is already useful. There are quite a few other things such as list of outstanding jobs, persisting across open/close of laptop/reboot for polling/monitoring, etc. I think we should separate some of these additional things into future PRs if we have time.

**Comment by psavery** [comment]
Done! Thanks Marcus!

**Comment by cryos** [comment]
There are a few minor failures due to clang-format issues that crept in, once fixed I will merge this in. Thanks Patrick.

**Comment by psavery** [comment]
Fixed.

**Pull request merged**